backend/server.py

```diff
@@ -226,43 +235,53 @@ def add_status():
     status_data = {}
     status_data['nodes'] = []
 
-
     return jsonify(return_data)
 
 @app.route('/fetcher/statistics', methods=['GET', 'POST'])
 def add_statistics():
+    timeout=30
+    waiter_statistics.close()
 
     global statistics_data
 
-    while statistics_data is None:
-        pass
+    if statistics_data is None:
+        reason=waiter_statistics.wait(timeout)
+        if reason==AwakeReason.timeout:
+            print('Timeout reached.')
+            waiter_statistics.close()
+            return jsonify({'Error': 'Timeout reached.'})
 
     time.sleep(1)  # Sleep for 1 second
-
+    waiter_statistics.close()
     return jsonify(statistics_data)
 
 
 @app.route('/fetcher/model', methods=['GET', 'POST'])
-def add_model():
-
+def add_model():                                       
+    timeout=30
+    global waiter_model
+    waiter_model.close()
     global statistics_server_id
     model_receiver_node.request_model(statistics_server_id)
-
     global model_data
-
-    while model_data is None:
-        pass
-
+    
+    reason=waiter_model.wait(timeout)
+    if reason==AwakeReason.timeout:
+        print(model_data)
+        print('Timeout reached.')
+        waiter_model.close()
+        return jsonify({'Error': 'Timeout reached.'})
+    
     time.sleep(1)  # Sleep for 1 second
 
     model = model_data
     model_data = None
-
+    waiter_model.close()
     return jsonify(model)
 
 @app.route('/train/<nJob>/<nIter>/<percentageData>', methods=['GET', 'POST'])
 def add_message(nJob, nIter, percentageData):
-
+    timeout=60
     global waiter_job
     waiter_job = IntWaitHandler(True)

```

Comment:
Since the `waiter_job` is being used globally, it would make more sense to declare it as a global variable like the other waiters, instead of initializing it within the function.